File: packages/engine/src/modules/cryo.ts
import { PHYSICAL_CONSTANTS } from "../constants";
import type {
  CryoControlMode,
  PolarProfileSummary,
  ResolvedStorageStream,
  SimParams,
  StorageInventory,
  Warning
} from "../types";

interface StreamProperties {
  densityKgPerM3: number;
  storageTemperatureK: number;
  latentHeatJPerKg: number;
  conditioningSecKWhPerKg: number;
  phaseLossEnabled: boolean;
}

export interface StorageDemand {
  id: string;
  stream: ResolvedStorageStream;
  role: StorageInventory["role"];
  rateKgPerDay: number;
}

interface PendingInventory extends StorageInventory {
  latentHeatJPerKg: number;
  phaseLossEnabled: boolean;
  areaM2: number;
  mliFluxWPerM2: number;
  hotSideTemperatureK: number;
}

export interface CryoOutput extends StreamProperties {
  stream: ResolvedStorageStream;
  controlMode: CryoControlMode;
  qLeakW: number;
  qRemovedW: number;
  qResidualW: number;
  unmitigatedBoiloffKgPerDay: number;
  boiloffKgPerDay: number;
  cryocoolerPowerW: number;
  mliFlux_WPerM2: number;
  cryoMassKg: number;
  totalReserveVolumeM3: number;
  totalConditioningPowerW: number;
  inventories: StorageInventory[];
  warnings: Warning[];
}

function propertiesFor(stream: ResolvedStorageStream, params: SimParams): StreamProperties {
  switch (stream) {
    case "lox":
      return { densityKgPerM3: 1141, storageTemperatureK: 90.2, latentHeatJPerKg: PHYSICAL_CONSTANTS.dHvap_LOX.value, conditioningSecKWhPerKg: 2.2, phaseLossEnabled: true };
    case "water-ice":
      return { densityKgPerM3: 917, storageTemperatureK: 150, latentHeatJPerKg: PHYSICAL_CONSTANTS.dHsub_ice.value, conditioningSecKWhPerKg: 0.15, phaseLossEnabled: true };
    case "liquid-water":
      return { densityKgPerM3: 997, storageTemperatureK: 293, latentHeatJPerKg: 0, conditioningSecKWhPerKg: 0.08, phaseLossEnabled: false };
    case "lh2":
      return { densityKgPerM3: 70.8, storageTemperatureK: 20.3, latentHeatJPerKg: PHYSICAL_CONSTANTS.dHvap_LH2.value, conditioningSecKWhPerKg: 12, phaseLossEnabled: true };
    case "lch4":
      return { densityKgPerM3: 422, storageTemperatureK: 111.7, latentHeatJPerKg: PHYSICAL_CONSTANTS.dHvap_LCH4.value, conditioningSecKWhPerKg: 1.2, phaseLossEnabled: true };
    case "co2-feed":
      return { densityKgPerM3: 1560, storageTemperatureK: 195, latentHeatJPerKg: PHYSICAL_CONSTANTS.dHsub_CO2.value, conditioningSecKWhPerKg: 0.15, phaseLossEnabled: true };
    case "custom":
      return { densityKgPerM3: params.rhoCryo, storageTemperatureK: params.Ttank, latentHeatJPerKg: params.customLatentHeatJPerKg, conditioningSecKWhPerKg: params.secLiquefaction, phaseLossEnabled: true };
  }
}

function pendingInventory(
  params: SimParams,
  demand: StorageDemand,
  profile: PolarProfileSummary
): PendingInventory {
  const properties = propertiesFor(demand.stream, params);
  const reserveInventoryKg = params.reserveDays * demand.rateKgPerDay;
  const volumeM3 = reserveInventoryKg / properties.densityKgPerM3;
  const radiusM = (3 * volumeM3 / (4 * Math.PI)) ** (1 / 3);
  const areaM2 = 4 * Math.PI * radiusM ** 2;
  const projectedAreaM2 = Math.PI * radiusM ** 2;
  const illumination = params.site === "polar" ? profile.averageIllumination : 1;
  const surfaceTemperatureK = params.site === "polar" ? profile.maximumSurfaceTemperatureK : params.Tsurface;
  const qSolarW = params.alphaTank * PHYSICAL_CONSTANTS.ISOLAR.value * projectedAreaM2 * illumination;
  const qAlbedoW = params.alphaTank * PHYSICAL_CONSTANTS.ISOLAR.value * 0.12 * params.Fview * areaM2 * illumination;
  const qIrW = params.epsTank * PHYSICAL_CONSTANTS.sigma.value * surfaceTemperatureK ** 4 * params.Fview * areaM2;
  const coldK = properties.storageTemperatureK;
  const qSpaceW = params.epsTank * PHYSICAL_CONSTANTS.sigma.value * (coldK ** 4 - 3 ** 4) * areaM2 * (1 - params.Fview);
  const qEnvironmentW = qSolarW + qAlbedoW + qIrW - qSpaceW;
  const equilibriumK = qEnvironmentW > 0
    ? (qEnvironmentW / (params.epsTank * PHYSICAL_CONSTANTS.sigma.value * areaM2)) ** 0.25
    : coldK;
  const hotK = Math.max(equilibriumK, coldK);
  const meanK = (hotK + coldK) / 2;

  // v0.1's coefficient is calibrated with layer density in layers/mm. Keep the
  // coefficient/unit pair intact until the open absolute heat-flux benchmark is accepted.
  const calibratedLayerDensity = params.Nlaydens / 10;
  const mliFluxWPerM2 =
    (params.C1mli * calibratedLayerDensity ** params.rExp * meanK * (hotK - coldK)) / params.Nmli +
    (params.C2mli * params.epsLayer * (hotK ** 4.67 - coldK ** 4.67)) / params.Nmli;
  const qLeakW = Math.max(0, mliFluxWPerM2 * areaM2) + params.qStrutW;
  const conditioningPowerW = demand.rateKgPerDay / 86_400 * properties.conditioningSecKWhPerKg * 3_600_000;

  return {
    id: demand.id,
    stream: demand.stream,
    role: demand.role,
    rateKgPerDay: demand.rateKgPerDay,
    reserveInventoryKg,
    volumeM3,
    storageMassKg: params.kCryoMass * demand.rateKgPerDay,
    densityKgPerM3: properties.densityKgPerM3,
    storageTemperatureK: properties.storageTemperatureK,
    conditioningSecKWhPerKg: properties.conditioningSecKWhPerKg,
    conditioningPowerW,
    qLeakW,
    qRemovedW: 0,
    qResidualW: qLeakW,
    unmitigatedLossKgPerDay: 0,
    actualLossKgPerDay: 0,
    latentHeatJPerKg: properties.latentHeatJPerKg,
    phaseLossEnabled: properties.phaseLossEnabled,
    areaM2,
    mliFluxWPerM2,
    hotSideTemperatureK: hotK
  };
}

export function simulateCryo(
  params: SimParams,
  demands: StorageDemand[],
  profile: PolarProfileSummary
): CryoOutput {
  const activeDemands = demands.filter((demand) => demand.rateKgPerDay > 1e-12);
  const fallback: StorageDemand = {
    id: "primary-product",
    stream: params.site === "equatorial" ? "lox" : "water-ice",
    role: "product",
    rateKgPerDay: params.targetKgPerDay
  };
  const pending = (activeDemands.length > 0 ? activeDemands : [fallback]).map((demand) => pendingInventory(params, demand, profile));
  const totalLeakW = pending.reduce((total, item) => total + item.qLeakW, 0);
  const removalBudgetW =
    params.cryoControlMode === "passive"
      ? 0
      : params.cryoControlMode === "capacity-limited"
        ? Math.min(totalLeakW, params.coolerCapacityW)
        : totalLeakW;
  let cryocoolerPowerW = 0;
  for (const item of pending) {
    item.qRemovedW = totalLeakW > 0 ? removalBudgetW * item.qLeakW / totalLeakW : 0;
    item.qResidualW = Math.max(0, item.qLeakW - item.qRemovedW);
    const lossFactor = item.phaseLossEnabled ? 86_400 / item.latentHeatJPerKg : 0;
    item.unmitigatedLossKgPerDay = item.qLeakW * lossFactor;
    item.actualLossKgPerDay = item.qResidualW * lossFactor;
    if (item.qRemovedW > 0 && item.hotSideTemperatureK > item.storageTemperatureK) {
      cryocoolerPowerW +=
        item.qRemovedW * (item.hotSideTemperatureK - item.storageTemperatureK) /
        (params.eta2ndLaw * item.storageTemperatureK);
    }
  }
  const inventories: StorageInventory[] = pending.map(({ latentHeatJPerKg: _latent, phaseLossEnabled: _phase, areaM2: _area, mliFluxWPerM2: _flux, hotSideTemperatureK: _hot, ...item }) => item);
  const primary = pending[0]!;
  const qRemovedW = pending.reduce((total, item) => total + item.qRemovedW, 0);
  const qResidualW = pending.reduce((total, item) => total + item.qResidualW, 0);
  const unmitigatedBoiloffKgPerDay = pending.reduce((total, item) => total + item.unmitigatedLossKgPerDay, 0);
  const boiloffKgPerDay = pending.reduce((total, item) => total + item.actualLossKgPerDay, 0);
  const cryoMassKg = pending.reduce((total, item) => total + item.storageMassKg, 0);
  const totalReserveVolumeM3 = pending.reduce((total, item) => total + item.volumeM3, 0);
  const totalConditioningPowerW = pending.reduce((total, item) => total + item.conditioningPowerW, 0);
  const totalAreaM2 = pending.reduce((total, item) => total + item.areaM2, 0);
  const warnings: Warning[] = [];
  if (params.cryoControlMode === "capacity-limited" && qResidualW > 1e-9) {
    warnings.push({ id: "cryo-capacity-shortfall", severity: "caution", module: "storage", message: "Shared cryocooler capacity is below the multi-inventory heat leak; residual phase loss remains.", value: qRemovedW, limit: totalLeakW });
  }

  return {
    stream: primary.stream,
    controlMode: params.cryoControlMode,
    densityKgPerM3: primary.densityKgPerM3,
    storageTemperatureK: primary.storageTemperatureK,
    latentHeatJPerKg: primary.latentHeatJPerKg,
    conditioningSecKWhPerKg: totalConditioningPowerW * 86_400 / (params.targetKgPerDay * 3_600_000),
    phaseLossEnabled: primary.phaseLossEnabled,
    qLeakW: totalLeakW,
    qRemovedW,
    qResidualW,
    unmitigatedBoiloffKgPerDay,
    boiloffKgPerDay,
    cryocoolerPowerW,
    mliFlux_WPerM2: totalAreaM2 > 0 ? pending.reduce((total, item) => total + item.mliFluxWPerM2 * item.areaM2, 0) / totalAreaM2 : 0,
    cryoMassKg,
    totalReserveVolumeM3,
    totalConditioningPowerW,
    inventories,
    warnings
  };
}

File: python/selene_isru/modules/cryo.py
from __future__ import annotations

import math
from typing import Any

from ..constants import c


def _properties(stream: str, params: dict[str, Any]) -> dict[str, Any]:
    values = {
        "lox": (1141, 90.2, c("dHvap_LOX"), 2.2, True),
        "water-ice": (917, 150, c("dHsub_ice"), 0.15, True),
        "liquid-water": (997, 293, 0, 0.08, False),
        "lh2": (70.8, 20.3, c("dHvap_LH2"), 12, True),
        "lch4": (422, 111.7, c("dHvap_LCH4"), 1.2, True),
        "co2-feed": (1560, 195, c("dHsub_CO2"), 0.15, True),
        "custom": (params["rhoCryo"], params["Ttank"], params["customLatentHeatJPerKg"], params["secLiquefaction"], True),
    }
    density, temperature, latent, conditioning, phase_loss = values[stream]
    return {"densityKgPerM3": density, "storageTemperatureK": temperature, "latentHeatJPerKg": latent, "conditioningSecKWhPerKg": conditioning, "phaseLossEnabled": phase_loss}


def _pending_inventory(params: dict[str, Any], demand: dict[str, Any], profile: dict[str, Any]) -> dict[str, Any]:
    properties = _properties(demand["stream"], params)
    reserve_inventory = params["reserveDays"] * demand["rateKgPerDay"]
    volume = reserve_inventory / properties["densityKgPerM3"]
    radius = (3 * volume / (4 * math.pi)) ** (1 / 3)
    area = 4 * math.pi * radius**2
    projected_area = math.pi * radius**2
    illumination = profile["averageIllumination"] if params["site"] == "polar" else 1
    surface_temperature = profile["maximumSurfaceTemperatureK"] if params["site"] == "polar" else params["Tsurface"]
    q_solar = params["alphaTank"] * c("ISOLAR") * projected_area * illumination
    q_albedo = params["alphaTank"] * c("ISOLAR") * 0.12 * params["Fview"] * area * illumination
    q_ir = params["epsTank"] * c("sigma") * surface_temperature**4 * params["Fview"] * area
    cold = properties["storageTemperatureK"]
    q_space = params["epsTank"] * c("sigma") * (cold**4 - 3**4) * area * (1 - params["Fview"])
    q_environment = q_solar + q_albedo + q_ir - q_space
    equilibrium = (q_environment / (params["epsTank"] * c("sigma") * area)) ** 0.25 if q_environment > 0 else cold
    hot = max(equilibrium, cold)
    mean = (hot + cold) / 2
    calibrated_layer_density = params["Nlaydens"] / 10
    mli_flux = (
        params["C1mli"] * calibrated_layer_density ** params["rExp"] * mean * (hot - cold) / params["Nmli"]
        + params["C2mli"] * params["epsLayer"] * (hot**4.67 - cold**4.67) / params["Nmli"]
    )
    q_leak = max(0, mli_flux * area) + params["qStrutW"]
    conditioning_power = demand["rateKgPerDay"] / 86400 * properties["conditioningSecKWhPerKg"] * 3_600_000
    return {
        "id": demand["id"], "stream": demand["stream"], "role": demand["role"], "rateKgPerDay": demand["rateKgPerDay"],
        "reserveInventoryKg": reserve_inventory, "volumeM3": volume, "storageMassKg": params["kCryoMass"] * demand["rateKgPerDay"],
        "densityKgPerM3": properties["densityKgPerM3"], "storageTemperatureK": cold,
        "conditioningSecKWhPerKg": properties["conditioningSecKWhPerKg"], "conditioningPowerW": conditioning_power,
        "qLeakW": q_leak, "qRemovedW": 0, "qResidualW": q_leak, "unmitigatedLossKgPerDay": 0, "actualLossKgPerDay": 0,
        "latentHeatJPerKg": properties["latentHeatJPerKg"], "phaseLossEnabled": properties["phaseLossEnabled"], "areaM2": area,
        "mliFluxWPerM2": mli_flux, "hotSideTemperatureK": hot,
    }


def simulate_cryo(params: dict[str, Any], demands: list[dict[str, Any]], profile: dict[str, Any]) -> dict[str, Any]:
    active = [demand for demand in demands if demand["rateKgPerDay"] > 1e-12]
    fallback = {"id": "primary-product", "stream": "lox" if params["site"] == "equatorial" else "water-ice", "role": "product", "rateKgPerDay": params["targetKgPerDay"]}
    pending = [_pending_inventory(params, demand, profile) for demand in (active or [fallback])]
    total_leak = sum(item["qLeakW"] for item in pending)
    if params["cryoControlMode"] == "passive":
        removal_budget = 0
    elif params["cryoControlMode"] == "capacity-limited":
        removal_budget = min(total_leak, params["coolerCapacityW"])
    else:
        removal_budget = total_leak
    cryocooler_power = 0.0
    for item in pending:
        item["qRemovedW"] = removal_budget * item["qLeakW"] / total_leak if total_leak > 0 else 0
        item["qResidualW"] = max(0, item["qLeakW"] - item["qRemovedW"])
        loss_factor = 86400 / item["latentHeatJPerKg"] if item["phaseLossEnabled"] else 0
        item["unmitigatedLossKgPerDay"] = item["qLeakW"] * loss_factor
        item["actualLossKgPerDay"] = item["qResidualW"] * loss_factor
        if item["qRemovedW"] > 0 and item["hotSideTemperatureK"] > item["storageTemperatureK"]:
            cryocooler_power += item["qRemovedW"] * (item["hotSideTemperatureK"] - item["storageTemperatureK"]) / (params["eta2ndLaw"] * item["storageTemperatureK"])
    public_keys = ("id", "stream", "role", "rateKgPerDay", "reserveInventoryKg", "volumeM3", "storageMassKg", "densityKgPerM3", "storageTemperatureK", "conditioningSecKWhPerKg", "conditioningPowerW", "qLeakW", "qRemovedW", "qResidualW", "unmitigatedLossKgPerDay", "actualLossKgPerDay")
    inventories = [{key: item[key] for key in public_keys} for item in pending]
    primary = pending[0]
    q_removed = sum(item["qRemovedW"] for item in pending)
    q_residual = sum(item["qResidualW"] for item in pending)
    unmitigated_loss = sum(item["unmitigatedLossKgPerDay"] for item in pending)
    actual_loss = sum(item["actualLossKgPerDay"] for item in pending)
    cryo_mass = sum(item["storageMassKg"] for item in pending)
    total_volume = sum(item["volumeM3"] for item in pending)
    total_conditioning_power = sum(item["conditioningPowerW"] for item in pending)
    total_area = sum(item["areaM2"] for item in pending)
    warnings: list[dict[str, Any]] = []
    if params["cryoControlMode"] == "capacity-limited" and q_residual > 1e-9:
        warnings.append({"id": "cryo-capacity-shortfall", "severity": "caution", "module": "storage", "message": "Shared cryocooler capacity is below the multi-inventory heat leak; residual phase loss remains.", "value": q_removed, "limit": total_leak})
    return {
        "stream": primary["stream"], "controlMode": params["cryoControlMode"], "densityKgPerM3": primary["densityKgPerM3"],
        "storageTemperatureK": primary["storageTemperatureK"], "latentHeatJPerKg": primary["latentHeatJPerKg"],
        "conditioningSecKWhPerKg": total_conditioning_power * 86400 / (params["targetKgPerDay"] * 3_600_000), "phaseLossEnabled": primary["phaseLossEnabled"],
        "qLeakW": total_leak, "qRemovedW": q_removed, "qResidualW": q_residual, "unmitigatedBoiloffKgPerDay": unmitigated_loss,
        "boiloffKgPerDay": actual_loss, "cryocoolerPowerW": cryocooler_power,
        "mliFlux_WPerM2": sum(item["mliFluxWPerM2"] * item["areaM2"] for item in pending) / total_area if total_area > 0 else 0,
        "cryoMassKg": cryo_mass, "totalReserveVolumeM3": total_volume, "totalConditioningPowerW": total_conditioning_power,
        "inventories": inventories, "warnings": warnings,
    }
